File: src/main/java/com/github/hirsivaja/ip/ipv4/Ipv4Packet.java
package com.github.hirsivaja.ip.ipv4;

import com.github.hirsivaja.ip.IpPacket;

import java.nio.ByteBuffer;

public record Ipv4Packet(Ipv4Header header, Ipv4Payload payload) implements IpPacket {

    @Override
    public void encode(ByteBuffer out) {
        header.encode(out);
        payload.encode(out);
    }

    @Override
    public int length() {
        return header.length() + payload.length();
    }

    public static IpPacket decode(ByteBuffer in) {
        return decode(in, true);
    }

    public static IpPacket decode(ByteBuffer in, boolean ensureChecksum) {
        Ipv4Header header = Ipv4Header.decode(in, ensureChecksum);
        byte[] payloadBytes = new byte[header.payloadLength()];
        in.get(payloadBytes);
        ByteBuffer payloadBuffer = ByteBuffer.wrap(payloadBytes);
        Ipv4Payload payload = Ipv4Payload.decode(payloadBuffer, ensureChecksum, header);
        return new Ipv4Packet(header, payload);
    }
}

File: src/main/kotlin/com/github/hirsivaja/ip/packet/Ipv4Packet.kt
package com.github.hirsivaja.ip.packet

import com.github.hirsivaja.ip.packet.header.Ipv4Header
import com.github.hirsivaja.ip.packet.payload.Ipv4Payload
import java.nio.ByteBuffer

data class Ipv4Packet(override val header : Ipv4Header, override val payload : Ipv4Payload) : IpPacket {
    companion object {
        fun decode(buf: ByteBuffer, ensureChecksum: Boolean): IpPacket {
            val header = Ipv4Header.decode(buf, ensureChecksum)
            val payload = Ipv4Payload.decode(buf, ensureChecksum, header)
            return Ipv4Packet(header, payload)
        }
    }
}
